reconciler/reconcile_common.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciler
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/logging"
+)
+
+const failedGenerationBump = "NewObservedGenFailure"
+
+// PreProcessReconcile contains logic to apply before reconciliation of a resource.
+func PreProcessReconcile(ctx context.Context, resource duckv1.KRShaped) {
+	condSet, err := getDefaultConditionSet(resource)
+	if err != nil {
+		return
+	}
+
+	newStatus := resource.GetStatus()
+	if newStatus.ObservedGeneration != resource.GetObjectMeta().GetGeneration() {
+		// Reset ready to unknown. The reconciler is expected to overwrite this.
+		manager := condSet.Manage(newStatus)
+		manager.MarkUnknown(
+			manager.GetTopLevelCondition().Type, failedGenerationBump, "unsucessfully observed a new generation")
+	}
```

Comment:
Don't we want to also call `InitializeConditions` here?

Reply:
My fear was that I don't necessarily want to clobber _all_ the conditions and I believe they should have been initialized in the admission webhook at this point?

Reply:
Controllers [call that method usually](https://github.com/knative/eventing/blob/e51b8a490e58762ce1cbda695785f277d6595294/pkg/client/injection/reconciler/sources/v1alpha2/containersource/stub/reconciler.go#L51-L52), webhooks only initialize the `spec` if I'm not mistaken (not every type has an admission webhook, for some third-party types the OpenAPI schema present in the CRD is sufficient).

I'm OK leaving that as it is, I was just curious what you thought about it.

Reply:
This is also not a complete condition set (we don't make any assumptions about the dependant sup-conditions here)

Reply:
hmm maybe we will want the full conditionset so that we can do this and at some point have the preprocessing do all the init stuff. I may reconsider this as another PR?